## 12 장           201803430 조성현


In [115]:
import pandas as pd
import numpy as np
import math
import scipy.stats as st
import warnings
import seaborn as sns; sns.set()
import matplotlib.pyplot as plt
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm
warnings.filterwarnings(action='ignore')

## 12.1

빈칸을 채워라

In [116]:
# (1) 계를 이용한 계산
print("(1)",round(26.02-18.46 , 4))

# (2) 자유도 합 계 이용
print("(2)",10-2)

# 평균제곱
#(3) = MSR =  SSR / df =2
MSR = 7.56 / 2
print("(3)",7.56 / 2)

#(4) = MSE = SSe / df = 8
MSE = 18.46 / 8
print("(4)",18.46 / 8)

# (5)  f값 f0 = MSR / MSE
print("(5)",round(MSR/MSE,5))

(1) 7.56
(2) 8
(3) 3.78
(4) 2.3075
(5) 1.63814


## 12.2
다음의 자료는 3종류의 필터를 조사하여 각각의 필터가 니코틴을 얼마나 걸러주는지 조사한 결과이다.

각 필터에 따라 성능에 차이가 있는지 가설을 설정하고 유의수준 5%로 검정하여라.

In [117]:
A = [5, 6, 8, 4, 6]
B = [6, 5, 4]
C =[7, 5, 6, 6]
#st.f_oneway(A,B,C) 내장 모듈

### 1:가설

H0: μA  =  μB  =  μC  (각 필터에 따라 성능에 차이가 없다)

H1: 적어도 하나의 μi는 다르다. (각 필터에 따라 성능에 차이가 있다)

In [118]:
all_data=A+B+C

x = np.array(A)
y = np.array(B)
z = np.array(C)
a_d = np.array(all_data)

ybar_x = x.mean()
ybar_y = y.mean()
ybar_z = z.mean()

ybar = a_d.mean()

k = 3

N = len(a_d)
n_x = len(x)
n_y = len(y)
n_z = len(z)

SSX = ((x - ybar_x)**2).sum()
SSY = ((y - ybar_y)**2).sum()
SSZ = ((z - ybar_z)**2).sum()

In [119]:
SSE = SSX+SSY+SSZ
SStr = n_x*((ybar_x - ybar)**2) + n_y*((ybar_y - ybar)**2) + n_z*((ybar_z - ybar)**2)
SST = SStr+SSE
MSE = SSE / (N-k)
MStr = SStr / (k-1)
f_0 = MStr / MSE

In [120]:
print("-----------------------------------------------------------------")
print("요인     |     제곱합      |      자유도    |     평균제곱   |     F값")
print("-----------------------------------------------------------------")
print("처리     |",round(SStr,3),"        |      ",k-1,"      | ",round(MStr,3),"     |  ",round(f_0 ,4))
print("-----------------------------------------------------------------")
print("잔차     |",round(SSE,3),"         |        ",N-k,"    | ",round(MSE,3),"     |    ")
print("-----------------------------------------------------------------")
print("계      |",round(SST,3),"        |       ",N-1,"   | ")
print("-----------------------------------------------------------------")



-----------------------------------------------------------------
요인     |     제곱합      |      자유도    |     평균제곱   |     F값
-----------------------------------------------------------------
처리     | 1.867         |       2       |  0.933      |   0.6562
-----------------------------------------------------------------
잔차     | 12.8          |         9     |  1.422      |    
-----------------------------------------------------------------
계      | 14.667         |        11    | 
-----------------------------------------------------------------


### 기각역 :

* F0 > F0.05 (2,9)  이면 H0 기각

처리와 잔차의 자유도를 가지는 F 분포 를 따른다.

In [121]:
a = 1-(0.05) # 단측 검정시 유의 수준 -> 0.005
f_a = st.f.ppf(a,k-1,N-k)
print(round(f_a,3))

4.256


귀무 가설 하에서 검정통계량 F 의 관측값 f0 = 0.6562 이므로

### 결론)

f0 < 4.256 즉 , 기각역에 속하지 않으므로 귀무가설 H0를 기각 할 수 없다. 

* 따라서 각 필터에 따라 성능 차이가 있다고 할 충분한 근거가 없다.

## 12.3
철선을 생산하는 공장에서 공정에 따른 철선의 인장강도에 차이가 있는지를 알아보기 위하여 네 종류의 공정에서 생산된 철선을 랜덤추출하여 표본을 취하여 인장강도를 측정하여 다음의 자료를 얻었다. 분산분석표를 작성하고, 공정에 따라 인장강도가 차이가 있다고 할 수 있는지 유의수준 5%에서 검정하여라.

In [122]:
A1 = [46.8, 58.0, 51.4, 61.0, 45.8]
A2 = [51.2, 62.4, 30.6, 46.0, 48.8]
A3 = [50.2, 38.2, 46.8, 26.7, 22.7]
A4 = [21.4, 22.1, 28.2, 42.7, 25.2]

df = pd.DataFrame({'A1': A1,'A2':A2,'A3':A3,'A4':A4})
df

A1    A2    A3    A4
0  46.8  51.2  50.2  21.4
1  58.0  62.4  38.2  22.1
2  51.4  30.6  46.8  28.2
3  61.0  46.0  26.7  42.7
4  45.8  48.8  22.7  25.2

### 1:가설

H0: μA1  =  μA2  =  μA3  = μA4    (인장강도에 차이가 없다)

H1: 적어도 하나의 μi는 다르다. (인장강도에 차이가 있다)

In [123]:
# 코드는 2 번문제와 매우 같다. A4의 인장강도에 관한것을 q 에 추가한다.
all_data=A1+A2+A3+A4

x = np.array(A1)
y = np.array(A2)
z = np.array(A3)
q = np.array(A4)

a_d = np.array(all_data)

ybar_x = x.mean()
ybar_y = y.mean()
ybar_z = z.mean()
ybar_q = q.mean()


ybar = a_d.mean()

k = 4

n = len(a_d)
n_x = len(x)
n_y = len(y)
n_z = len(z)
n_q = len(q)

SSX = ((x - ybar_x)**2).sum()
SSY = ((y - ybar_y)**2).sum()
SSZ = ((z - ybar_z)**2).sum()
SSq = ((q - ybar_q)**2).sum()

In [124]:
SSE = SSX+SSY+SSZ+SSq
SStr = n_x*((ybar_x - ybar)**2) + n_y*((ybar_y - ybar)**2) + n_z*((ybar_z - ybar)**2)+ n_q*((ybar_q - ybar)**2)
SST = SStr+SSE
MSE = SSE / (n-k)
MStr = SStr / (k-1)
f_0 = MStr / MSE # 검정통계량

* 분산분석 표및 검정통계량의 관측값 F0는 다음과 같다.

In [125]:
print("-----------------------------------------------------------------")
print("요인     |     제곱합      |      자유도    |     평균제곱   |     F값")
print("-----------------------------------------------------------------")
print("처리     |",round(SStr,3),"     |      ",k-1,"      | ",round(MStr,3),"   |  ",round(f_0 ,4))
print("-----------------------------------------------------------------")
print("잔차     |",round(SSE,3),"     |     ",n-k,"      | ",round(MSE,3),"    |    ")
print("-----------------------------------------------------------------")
print("계      |",round(SST,3),"     |     ",n-1,"       | ")
print("-----------------------------------------------------------------")



-----------------------------------------------------------------
요인     |     제곱합      |      자유도    |     평균제곱   |     F값
-----------------------------------------------------------------
처리     | 1840.742      |       3       |  613.581    |   6.1728
-----------------------------------------------------------------
잔차     | 1590.416      |      16       |  99.401     |    
-----------------------------------------------------------------
계      | 3431.158      |      19        | 
-----------------------------------------------------------------


귀무 가설 하에서 검정통계량 F 의 관측값 f0 = 6.1762 이다.

### 기각역 :
귀무 가설하 검정통계량은 처리와 잔차의 자유도를 가지는 F 분포 를 따르므로,
* F0 > F0.05 (3,16)  이면 H0 기각



In [126]:
a = 1-(0.05) # 단측 검정시 유의 수준 -> 0.005
f_a = st.f.ppf(a,k-1,N-k)
print(" F 0.05(3,16):",round(f_a,3))

 F 0.05(3,16): 4.066


### 결론)

f0 > 4.066 즉 , 기각역에 속하므로 귀무가설 H0를 유의수준 5% 하에서 기각 할 수 있다. 

* 따라서 각 인장강도간 차이가 있다고 할 충분한 근거가 있다.

## 12.5
어떤 영양제에 첨가된 요소들이 모두 비슷한 작용을 한다는 사실을 확인하기 위해 쥐에게 그 영양제를 주입하여 몸무게를 측정한 자료이다. 

그 영양제가 효과가 있는지 유의수준 5%에서 검정하여라.

In [127]:
A = [110, 122, 132, 114, 105]
B = [122, 131, 114, 108, 110]
C = [110, 120, 112, 108, 100]

df = pd.DataFrame({'A': A,'B':B,'C':C})
df

A    B    C
0  110  122  110
1  122  131  120
2  132  114  112
3  114  108  108
4  105  110  100

### 1:가설

H0: μA  =  μB  =  μC   (영양제 별로 효과의 차이가 없다)

H1: 적어도 하나의 μi는 다르다. (영양제 별로 효과의 차이가있다)

In [128]:
all_data=A+B+C

x = np.array(A)
y = np.array(B)
z = np.array(C)
a_d = np.array(all_data)

ybar_x = x.mean()
ybar_y = y.mean()
ybar_z = z.mean()

ybar = a_d.mean()

k = 3

n = len(a_d)
n_x = len(x)
n_y = len(y)
n_z = len(z)

SSX = ((x - ybar_x)**2).sum()
SSY = ((y - ybar_y)**2).sum()
SSZ = ((z - ybar_z)**2).sum()

In [129]:
SSE = SSX+SSY+SSZ
SStr = n_x*((ybar_x - ybar)**2) + n_y*((ybar_y - ybar)**2) + n_z*((ybar_z - ybar)**2)
SST = SStr+SSE
MSE = SSE / (n-k)
MStr = SStr / (k-1)
f_0 = MStr / MSE

In [130]:
print("-----------------------------------------------------------------")
print("요인     |     제곱합      |      자유도    |     평균제곱   |     F값")
print("-----------------------------------------------------------------")
print("처리     |",round(SStr,3),"        |      ",k-1,"    | ",round(MStr,3),"     |  ",round(f_0 ,4))
print("-----------------------------------------------------------------")
print("잔차     |",round(SSE,3),"         |      ",n-k,"   | ",round(MSE,3),"     |    ")
print("-----------------------------------------------------------------")
print("계      |",round(SST,3),"        |    ",n-1,"     | ")
print("-----------------------------------------------------------------")



-----------------------------------------------------------------
요인     |     제곱합      |      자유도    |     평균제곱   |     F값
-----------------------------------------------------------------
처리     | 154.533         |       2     |  77.267      |   0.9097
-----------------------------------------------------------------
잔차     | 1019.2          |       12    |  84.933      |    
-----------------------------------------------------------------
계      | 1173.733         |     14      | 
-----------------------------------------------------------------


귀무 가설 하에서 검정통계량 F 의 관측값 f0 = 0.9097 이다.

### 기각역 :

귀무 가설하 검정통계량은 처리와 잔차의 자유도를 가지는 F 분포 를 따르므로,
* F0 > F0.05 (2,12)  이면 H0 기각



In [131]:
a = 1-(0.05) # 단측 검정시 유의 수준 -> 0.005
f_a = st.f.ppf(a,k-1,N-k)
print(" F 0.05(2,12):",round(f_a,3))

 F 0.05(2,12): 4.256


### 결론)

f0 < 3.885 즉 , 기각역에 속하지 않으므로 귀무가설 H0를 유의수준 5% 하에서 기각 할 수 없다. 

* 따라서 각 영양제에 첨가된 요소들이 차이가 있다고 할 충분한 근거가 없다라고 볼 수 있다.

## 12.10
A회사에서 급여 지급 방법이 판매량에 영향을 주는지 알아보고자 한다. 

세 가지(고정월급, 판매실적, 고정월급 + 판매실적)의 급여 방법에 대해 랜덤하게 뽑은 판매원의 판매실적 결과가 다음과 같다.

In [132]:

A = [42, 45, 44, 43, 44,None]
B = [43, 45, 51, 48, 47, 49]
C = [44, 49, 47, 50,None,None]

pay = pd.DataFrame({'Sallary': A,'Selling':B,'Sallary&Selling':C})
pay

Sallary  Selling  Sallary&Selling
0     42.0       43             44.0
1     45.0       45             49.0
2     44.0       51             47.0
3     43.0       48             50.0
4     44.0       47              NaN
5      NaN       49              NaN

### SST, SSE, SStr 값을 구하여라.

In [133]:
A = [42, 45, 44, 43, 44]
B = [43, 45, 51, 48, 47, 49]
C = [44, 49, 47, 50]

all_data=A+B+C

x = np.array(A)
y = np.array(B)
z = np.array(C)
a_d = np.array(all_data)

ybar_x = x.mean()
ybar_y = y.mean()
ybar_z = z.mean()

ybar = a_d.mean()

k = 3

n = len(a_d)
n_x = len(x)
n_y = len(y)
n_z = len(z)

SSX = ((x - ybar_x)**2).sum()
SSY = ((y - ybar_y)**2).sum()
SSZ = ((z - ybar_z)**2).sum()
SSE = SSX+SSY+SSZ
SStr = n_x*((ybar_x - ybar)**2) + n_y*((ybar_y - ybar)**2) + n_z*((ybar_z - ybar)**2)
SST = SStr+SSE


In [134]:
print("SST=",round(SST,3)," ","SSE=",round(SSE,3)," ","SStr=",round(SStr,3))

SST= 112.933   SSE= 67.033   SStr= 45.9


### F값을 구하고, 급여 방법에 따라 판매실적에 차이가 있는지 유의수준 1%에서 검정하여라. 


In [135]:
MSE = SSE / (n-k)
MStr = SStr / (k-1)
f_0 = MStr / MSE
print("F값 = ",round(f_0,4))

F값 =  4.1084


### 기각역 :

귀무 가설하 검정통계량의 관측값은 처리와 잔차의 자유도를 가지는 F 분포 를 따르므로,
* F0 > F0.01 (2,12)  이면 H0 기각

In [136]:
a = 1-(0.01) # 단측 검정시 유의 수준 -> 0.005
f_a = st.f.ppf(a,k-1,N-k)
print(" F 0.01(2,12):",round(f_a,3))

 F 0.01(2,12): 8.022


### 결론)

f0 < 6.927 즉 , 귀무 가설하 검정통계량의 관측값이 기각역에 속하지 않으므로 귀무가설 H0를 유의수준 1% 하에서 기각 할 수 없다. 

* 따라서 각 급여방법에 따라 판매원의 판매실적 차이가 있다고 할 충분한 근거가 없다라고 볼 수 있다.